cfgov/v1/models/banners.py

```diff
@@ -0,0 +1,38 @@
+from django.core.validators import RegexValidator
+from django.db import models
+from django.utils.safestring import mark_safe
+
+from wagtail.admin.edit_handlers import FieldPanel, StreamFieldPanel
+from wagtail.core.fields import StreamField
+
+from v1.atomic_elements.molecules import Notification
+
+
+class Banner(models.Model):
+    title = models.CharField(
+        max_length=255,
+        help_text='For internal reference only; does not appear on the site.'
+    )
+    url_patterns = models.CharField(
+        max_length=1000,
+        verbose_name='URL patterns',
+        help_text=mark_safe('A regular expression pattern for matching URLs '
+                            'that should show the banner, for example: '
+                            '<code>contact-us|^/complaint/$</code>'),
+        validators=[
+            RegexValidator(regex=r'[A-Za-z0-9\-_.:/?&|\^$]')
+        ],
+    )
+    # TODO: add `min_num` and `max_num` arguments of 1 to the StreamField below
+    # if https://github.com/wagtail/wagtail/pull/5185 ever gets merged.
+    content = StreamField([
+        ('content', Notification()),
+    ])
+    enabled = models.BooleanField()
+
+    panels = [
+        FieldPanel('title'),
+        FieldPanel('url_patterns'),
```

Comment:
Worth using a [RegexField](https://docs.djangoproject.com/en/2.2/ref/forms/fields/#regexfield) in the editor?

```py
FieldPanel('url_patterns', widget=forms.RegexField(regex=REGEX_PATTERN))
```

Reply:
🤔 I'm not sure. Didn't know that was an option. I'll try it out.

Reply:
Doesn't work.

```
File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/exception.py" in inner
  41.             response = get_response(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _legacy_get_response
  249.             response = self._get_response(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  57.         response = view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/urls/__init__.py" in wrapper
  102.             return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/decorators.py" in decorated_view
  34.             return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/options.py" in create_view
  326.         return view_class.as_view(**kwargs)(request)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in view
  68.             return self.dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in dispatch
  691.         return super().dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/utils/decorators.py" in _wrapper
  67.             return bound_func(*args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  23.                 return view_func(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/utils/decorators.py" in bound_func
  63.                 return func.__get__(self, type(self))(*args2, **kwargs2)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in dispatch
  79.         return super().dispatch(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in dispatch
  88.         return handler(request, *args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/base.py" in get
  155.         context = self.get_context_data(**kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in get_context_data
  150.         form = self.get_form()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/views/generic/edit.py" in get_form
  44.             form_class = self.get_form_class()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/contrib/modeladmin/views.py" in get_form_class
  127.         return self.get_edit_handler().get_form_class()

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/edit_handlers.py" in get_form_class
  344.             widgets=self.widget_overrides())

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/edit_handlers.py" in get_form_for_model
  62.     return metaclass(class_name, (form_class,), form_class_attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/forms/models.py" in __new__
  66.         new_class = super(WagtailAdminModelFormMetaclass, cls).__new__(cls, name, bases, attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/modelcluster/forms.py" in __new__
  245.         new_class = super(ClusterFormMetaclass, cls).__new__(cls, name, bases, attrs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/models.py" in __new__
  266.                 apply_limit_choices_to=False,

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/models.py" in fields_for_model
  186.             formfield = formfield_callback(f, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/wagtail/admin/forms/models.py" in formfield_for_dbfield
  47.     return db_field.formfield(**kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/db/models/fields/__init__.py" in formfield
  1110.         return super(CharField, self).formfield(**defaults)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/db/models/fields/__init__.py" in formfield
  891.         return form_class(**defaults)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in __init__
  228.         super(CharField, self).__init__(*args, **kwargs)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in __init__
  109.         extra_attrs = self.widget_attrs(widget)

File "/Users/cranfills/.virtualenvs/cfgov-refresh/lib/python3.6/site-packages/django/forms/fields.py" in widget_attrs
  246.         if self.max_length is not None and not widget.is_hidden:

Exception Type: AttributeError at /admin/v1/banner/create/
Exception Value: 'RegexField' object has no attribute 'is_hidden'
```

Reply:
You're right, I was mixing up widgets and form fields.